src/librustc_typeck/astconv.rs

```diff
@@ -2504,18 +2534,51 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         path_segs
     }
 
+    fn lint_type_param_on_variant_ctor(&self, segments: &[hir::PathSegment<'_>]) {
+        // Doing this to get around rustfmt-caused line too long.
+        use hir::PathSegment as P;
+        if let [.., prev, P { ident, hir_id: Some(hir_id), args: Some(args), .. }] = segments {
+            let sp = args.span;
+            if sp.hi() == sp.lo() || sp == DUMMY_SP || sp.parent().is_some() {
+                // The params were not written by the user, but rather derived. These are expected.
+                // `Enum::Variant` where `Variant` has inferred lifetimes.
+                return;
+            }
```

Comment:
I don't like this check, and you can get rid of the lifetimes for now (they shouldn't be added to value paths anyway).

Reply:
Agree. It's hacky as hell.

> you can get rid of the lifetimes for now

How?

> they shouldn't be added to value paths anyway

I've spent too much time trying to figure out of a way of making this change without affecting currently accepted code.

Reply:
Basically you want all `Expr::Struct` with variant paths to behave in HRI lowering like they did when you had removed the `DefKind::Variant` arm.
You don't want the elision lifetimes being injected *at all*.

They're for `fn foo(x: &T) -> Enum::Variant`, or perhaps `<Enum>::Variant {...}` (neither of which compile right now), *not* `Enum::Variant {...}`.

Reply:
But if none of the PathSegments _have_ the elision lifetimes injected (in src/librustc_ast_lowering/path.rs) then you get knock down errors about the type for a HirId not existing .

Reply:
> then you get knock down errors about the type for a HirId not existing

I don't recall you bringing this up, I thought it just worked while you had that arm removed? I don't see why it would, other value paths have no lifetimes injected into them *at all*.